[WTF] Leave top granule of SequesteredArenaAllocator behind

Achierius opened the pull request:
#### da4298fbcd76378ad2d2b03a8212ea79854ec94f
<pre>
[WTF] Leave top granule of SequesteredArenaAllocator behind
<a href="https://bugs.webkit.org/show_bug.cgi?id=288727">https://bugs.webkit.org/show_bug.cgi?id=288727</a>
<a href="https://rdar.apple.com/145760952">rdar://145760952</a>

Reviewed by Yusuke Suzuki and Yijia Huang.

This is a neat and simple version of a thread-local cache. Since we know
that all objects in the granule will have been freed, we can freely
re-use it for the next allocation cycle -- thereby avoiding the initial
slow-path allocation for the first allocation of a compilation job.
Ideally, the granule size will be large enough that this will suffice
for most jobs.

* Source/WTF/wtf/DoublyLinkedList.h:
(WTF::DoublyLinkedList&lt;T&gt;::splitAt):
* Source/WTF/wtf/SequesteredAllocator.h:

Canonical link: <a href="https://commits.webkit.org/291255@main">https://commits.webkit.org/291255@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d4901644f5a768fe133fb4100b878e3df508779b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/92445 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/11987 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/1587 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/97429 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/42952 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/12264 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/20447 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/70842 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/42952 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/95447 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/9327 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/83723 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51172 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/9024 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/1360 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/42283 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/85155 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/79349 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/1301 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/99455 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/91111 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/19496 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/14399 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/79859 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/19746 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/79603 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/79132 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/23659 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/951 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/12505 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14716 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/19480 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/24652 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/113759 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/19167 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/32897 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/22627 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/20907 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hyjorc1 commented on Source/WTF/wtf/DoublyLinkedList.h:
```diff
@@ -160,6 +162,28 @@ template<typename T> inline void DoublyLinkedList<T>::append(T* node)
+    newList.m_tail = m_tail;
+    m_tail = p;
+    newList.m_head->setPrev(0);
+    m_tail->setNext(0);
```
nit: better logic flow?
```
DoublyLinkedList<T> newList { };
newList.m_head = p->next();
newList.m_head->setPrev(0);
newList.m_tail = m_tail;

m_tail = p;
m_tail->setNext(0);
```

hyjorc1 commented on Source/WTF/wtf/DoublyLinkedList.h:
```diff
@@ -160,6 +162,28 @@ template<typename T> inline void DoublyLinkedList<T>::append(T* node)
     m_tail = node;
 }
 
+template<typename T> inline DoublyLinkedList<T> DoublyLinkedList<T>::splitAt(int toKeep)
```
`int toKeep` should we use `int`? Then what happens if `toKeep` is a negative value?

hyjorc1 commented on Source/WTF/wtf/DoublyLinkedList.h:
```diff
@@ -160,6 +162,29 @@ template<typename T> inline void DoublyLinkedList<T>::append(T* node)
+
+    if (!p || !p->next())
+        return { };
+    for (auto i = 1; i < toKeep; i++) {
```
Should we just use `size_t` for `i`?

hyjorc1 reviewed (approved):
r=me

Constellation reviewed (approved):
Let's write a test in TestWTF.

Achierius committed:
[WTF] Leave top granule of SequesteredArenaAllocator behind
https://bugs.webkit.org/show_bug.cgi?id=288727
rdar://145760952

Reviewed by Yusuke Suzuki and Yijia Huang.

This is a neat and simple version of a thread-local cache. Since we know
that all objects in the granule will have been freed, we can freely
re-use it for the next allocation cycle -- thereby avoiding the initial
slow-path allocation for the first allocation of a compilation job.
Ideally, the granule size will be large enough that this will suffice
for most jobs.

* Source/WTF/wtf/DoublyLinkedList.h:
(WTF::DoublyLinkedList<T>::splitAt):
* Source/WTF/wtf/SequesteredAllocator.h:

Canonical link: https://commits.webkit.org/291255@main
